File: java/SinglePileNIM.java
/* 
 * Single Pile NIM - Java Edition
 * By: M.S.T.O.P.
 * March 2017
 */

import java.util.Scanner;
import java.util.Random;
import java.lang.Math;

public class SinglePileNIM 
{
	public static void main(String[] args)
	{
		// Init Variables
		int currentNum;
		Random ran = new Random();
		
		int takeAmount = 0;
		int maxTake;
		String answer;
		
		int whoseTurn = 1;
		boolean gameIsOver = false;
		boolean gameLoop = true;
		boolean validInput;
		
		Scanner input = new Scanner(System.in);
		
		// Main Game Loop
		while (gameLoop)
		{
			// Setup
			maxTake = ran.nextInt(4) + 2;
			currentNum = maxTake * (ran.nextInt(4) + 3) + ran.nextInt(maxTake) + 1;
			gameIsOver = false;
			
			System.out.println("Single Pile NIM\n---------------\n");
			
			// Session loop
			while (!gameIsOver)
			{
				// Player's turn
				validInput = false;
				whoseTurn = 1;
				
				while (!validInput)
				{
					// Get an integer
					while (true)
					{
						System.out.println("Remaining total is: " + currentNum);
						System.out.println("How much will you take (1-" + maxTake + ")?");
						
						try
						{
							takeAmount = Integer.parseInt(input.nextLine());
							break;
						}
						
						catch (Exception e)
						{
							System.out.println("Invalid amount!\n");							
						}
					}					
						
					// Check to see if input is within the valid range
					if (takeAmount > 0 && takeAmount <= maxTake)
						validInput = true;
				
					else
						System.out.println("Invalid amount!\n");
				}
				
				System.out.println("You took " + takeAmount + ".\n");
				currentNum -= takeAmount;
				
				if (currentNum <= 0)
					gameIsOver = true;
				
				else
				{
					// Computer's turn
					whoseTurn = 2;
					
					takeAmount = Math.max(1,((currentNum % (maxTake+1)) + maxTake) % (maxTake+1));
					System.out.println("I will take " + takeAmount + ".\n");
					
					currentNum -= takeAmount;
					
					if (currentNum <= 0)
						gameIsOver = true;
				}
			}
			
			// Determine the winner
			if (whoseTurn == 1)
				System.out.println("I win, You lose.");
			
			else
				System.out.println("You win, I lose.");
			
			//Play again?
			validInput = false;
			
			// Get an string
			while (!validInput)
			{
				System.out.println("Would you like to play again? (Y/N)");
				answer = input.next();
				
				answer = answer.toUpperCase();
				
				// Check to see if input is "yes" or "no"
				if (answer.equals("N") || answer.equals("NO"))
				{
					gameLoop = false;
					validInput = true;
				}
				
				else if (answer.equals("Y") || answer.equals("YES"))
				{
					System.out.println("");
					validInput = true;
				}
				
				else
					System.out.println("Invalid answer!\n");
			}
		}
		
		input.close();
		System.out.println("\nGoodbye!");
	}
}

File: ruby/SinglePileNIM.rb
#
# Single Pile NIM - Ruby Edition
# By: M.S.T.O.P.
# August 2017
#

# Init variables
current_num = 0

take_amount = 0
max_take = 0
answer = ""

whose_turn = 1
game_is_over = false
game_loop = true

# Main Game Loop
while game_loop do

  # Setup
  max_take = 2 + rand(4)
  current_num = max_take * (3 + rand(4)) + 1 + rand(max_take)
  game_is_over = false

  puts "Single Pile NIM\n---------------\n"

  # Session loop
  while !game_is_over do

    # Get an integer within range
    begin
      puts "\nRemaining total is: #{current_num}"
      puts "How much will you take (1-#{max_take})?"
      take_amount = Integer(gets.chomp)
      raise if take_amount < 1 || take_amount > max_take
    rescue
      puts "Invalid amount!\n"
      retry
    end

    puts "You took #{take_amount}.\n"
    current_num -= take_amount

    if current_num <= 0
      game_is_over = true
    else
      # Computer's turn
      whose_turn = 2

      take_amount = [ 1, ((current_num % ( max_take + 1 )) + max_take ) % ( max_take + 1 ) ].max
      puts "I will take #{take_amount}.\n"

      current_num -= take_amount

      game_is_over = true if current_num <= 0
    end
  end

  # Determine the winner
  if whose_turn == 1
    puts "\nI win, you lose."
  else
    puts "\nYou win, I lose."
  end

  # Play again?
  begin
    puts "\nWould you like to play again? (Y/N)"
    answer = gets.chomp.upcase

    if answer == "N" || answer == "NO"
      game_loop = false
    elsif answer == "Y" || answer == "YES"
      puts ""
    else
      raise
    end
  rescue
    puts "Invalid answer!\n"
    retry
  end
end

puts "\nGoodbye!"